internal/sponge/item/item.go

```diff
@@ -0,0 +1,282 @@
+			return err
+		}
+
+		new = true
+	}
+
+	// Insert or update the item.
+	f := func(c *mgo.Collection) error {
+		q := bson.M{"_id": item.Id}
+		log.Dev(context, "Upsert", "MGO : db.%s.upsert(%s, %s)", c.Name, mongo.Query(q), mongo.Query(item))
+		_, err := c.Upsert(q, item)
+		return err
+	}
+
+	if err := db.ExecuteMGO(context, Collection, f); err != nil {
+		log.Error(context, "Upsert", err, "Completed")
+		return err
+	}
+
+	if new {
+		// historical code
+	}
+
+	// if the item isn't new it may be in various caches
+	//   flush the whole cache
+	if !new {
+		cache.Flush()
+	}
+
+	return nil
+}
+
+// GetById retrieves an item by its id, given the id as a string.
+func GetByIdString(context interface{}, db *db.DB, i string) (*Item, error) {
+
+	// can we make this into a valid bson ObjectId?
+	id := bson.ObjectIdHex(i)
+	//	if err != nil {
+	//		return nil, err
+	//	}
+
+	// if so, use the traditional GetById
+	return GetById(context, db, id)
+}
+
+// GetById retrieves an item by its id.
+func GetById(context interface{}, db *db.DB, id bson.ObjectId) (*Item, error) {
+	log.Dev(context, "GetById", "Started : Id[%s]", id.Hex())
+
+	var item Item
+
+	// check if the item is in the cache
+	key := "item-" + id.Hex()
+	if v, found := cache.Get(key); found {
+		item := v.(Item)
+		log.Dev(context, "GetById", "Completed : CACHE : Item[%+v]", &item)
+		return &item, nil
+	}
+
+	// query the database for the item
+	f := func(c *mgo.Collection) error {
+		q := bson.M{"_id": id}
+		log.Dev(context, "GetById", "MGO : db.%s.findOne(%s)", c.Name, mongo.Query(q))
+		return c.Find(q).One(&item)
+	}
+
+	if err := db.ExecuteMGO(context, Collection, f); err != nil {
+		if err == mgo.ErrNotFound {
+			err = ErrNotFound
+		}
+
+		log.Error(context, "GetById", err, "Completed")
+		return nil, err
+	}
+
+	// set the cache: TODO, caching based on type params
+	cache.Set(key, item, gc.DefaultExpiration)
+
+	log.Dev(context, "GetById", "Completed : Item[%+v]", &item)
+	return &item, nil
+}
+
+// GetById retrieves items by an array of ids
+func GetByIds(context interface{}, db *db.DB, ids []bson.ObjectId) (*[]Item, error) {
+	log.Dev(context, "GetByIds", "Started : Looking for %s ids", len(ids))
+
+	var items []Item
+
+	// query the database for the item
+	f := func(c *mgo.Collection) error {
+		q := bson.M{"_id": bson.M{"$in": ids}}
+		log.Dev(context, "GetByIds", "MGO : ", c.Name, mongo.Query(q))
+		return c.Find(q).All(&items)
+	}
+
+	if err := db.ExecuteMGO(context, Collection, f); err != nil {
+		if err == mgo.ErrNotFound {
+			err = ErrNotFound
+		}
+
+		log.Error(context, "GetByIds", err, "Completed")
+		return nil, err
+	}
+
+	log.Dev(context, "GetByIds", "Completed : Found %+v items", len(items))
+	return &items, nil
+}
+
+// GetByQuery accepts a bson.M query and runs it against the item collection
+//  caution should be used to only query against indexed fields
+func GetByQuery(context interface{}, db *db.DB, q bson.M) (*[]Item, error) {
```

Comment:
We shouldn't have to do this. This forces the user to import mgo, which should never be.

Rule of thumb:
If we need to do something once, do it.
Twice, copy and paste
Three time, then abstract it.